executor: add 'racedebug' option for syz-execprog

a13xp0p0v opened the pull request:
Sometimes race conditions are reproduced by syz-execprog and are not
reproduced by the programs generated with syz-prog2c. In such cases
it's very helpful to know when exactly the fuzzing syscalls are executed.

Unfortunately, adding timestamps to the output of the original 'debug'
option doesn't work. This option provides very verbose output, which
slows down executor and breaks the repro.

So let's add the 'racedebug' option for syz-execprog, which provides the
limited debug output with timestamps from executor.

Signed-off-by: Alexander Popov <alex.popov@linux.com>

*******************************************************************************
Before sending a pull request, please review Contribution Guidelines:
https://github.com/google/syzkaller/blob/master/docs/contributing.md
*******************************************************************************

dvyukov commented:
> Excuse me, I don't see the reason why travis-ci continuous integration test is failed...

It says the code is not formatted.
I can reproduce the same locally:

```
syzkaller$ make format
syzkaller$ git diff
diff --git a/executor/executor.cc b/executor/executor.cc
index a3fda45f..2c93c124 100644
--- a/executor/executor.cc
+++ b/executor/executor.cc
@@ -476,9 +476,9 @@ void receive_execute()
        if (!flag_threaded)
                flag_collide = false;
        race_debug("exec opts: pid=%llu threaded=%d collide=%d cover=%d comps=%d dedup=%d fault=%d/%d/%d prog=%llu\n",
-             procid, flag_threaded, flag_collide, flag_collect_cover, flag_collect_comps,
-             flag_dedup_cover, flag_inject_fault, flag_fault_call, flag_fault_nth,
-             req.prog_size);
+                  procid, flag_threaded, flag_collide, flag_collect_cover, flag_collect_comps,
+                  flag_dedup_cover, flag_inject_fault, flag_fault_call, flag_fault_nth,
+                  req.prog_size);
        if (SYZ_EXECUTOR_USES_SHMEM) {
                if (req.prog_size)
                        fail("need_prog: no program");
@@ -1326,4 +1326,3 @@ void race_debug(const char* msg, ...)
        va_end(args);
        fflush(stderr);
 }
-
diff --git a/pkg/ipc/ipcconfig/ipcconfig.go b/pkg/ipc/ipcconfig/ipcconfig.go
index 0a50993b..af7e9444 100644
--- a/pkg/ipc/ipcconfig/ipcconfig.go
+++ b/pkg/ipc/ipcconfig/ipcconfig.go
@@ -13,13 +13,13 @@ import (
 )
 
 var (
-       flagExecutor = flag.String("executor", "./syz-executor", "path to executor binary")
-       flagThreaded = flag.Bool("threaded", true, "use threaded mode in executor")
-       flagCollide  = flag.Bool("collide", true, "collide syscalls to provoke data races")
-       flagSignal   = flag.Bool("cover", false, "collect feedback signals (coverage)")
-       flagSandbox  = flag.String("sandbox", "none", "sandbox for fuzzing (none/setuid/namespace)")
-       flagDebug    = flag.Bool("debug", false, "debug output from executor")
-       flagTimeout  = flag.Duration("timeout", 0, "execution timeout")
+       flagExecutor  = flag.String("executor", "./syz-executor", "path to executor binary")
+       flagThreaded  = flag.Bool("threaded", true, "use threaded mode in executor")
+       flagCollide   = flag.Bool("collide", true, "collide syscalls to provoke data races")
+       flagSignal    = flag.Bool("cover", false, "collect feedback signals (coverage)")
+       flagSandbox   = flag.String("sandbox", "none", "sandbox for fuzzing (none/setuid/namespace)")
+       flagDebug     = flag.Bool("debug", false, "debug output from executor")
+       flagTimeout   = flag.Duration("timeout", 0, "execution timeout")
        flagRaceDebug = flag.Bool("racedebug", false, "limited debug output with timestamps from executor")
 )
```

dvyukov commented:
> By the way, I've just noticed that some recent changes in syzkaller are not backward-compatible.

High number of changes in syzkaller throughout its history is not backward-compatible. It's practically impossible to preserve all tiny variations that existed throughout the history. But git is very good at it ;)

a13xp0p0v commented:
> It says the code is not formatted.
Thanks, I've fixed the code style and pushed it again.

And does this feature look useful to you?

dvyukov commented:
> And does this feature look useful to you?

It looks useful to me, but not as a separate feature.
What exactly we print in debug mode is poorly defined, what we print in this mode is also not completely defined (e.g. why exec opts, but not test process pid?). I hate to introduce more flags for rarely-used features used by few people from time to time. They will break.

What about tidying up the current output to the form that will serve your need too?

I've captured the current debug output on a simple program:

```
exec opts: pid=0 threaded=1 collide=1 cover=1 comps=0 dedup=1 fault=0/0/0 prog=0
spawned worker pid 238
scheduling call 0 [getpid] on thread 0
[   49655] #0: getpid()
#0: read cover size = 38
#0: getpid = 0xee
completion of call 0 [getpid] on thread 0
out #0: index=0 num=199 errno=0 finished=1 blocked=0 sig=38 cover=38 comps=0
scheduling call 1 [getpid] on thread 0
[   49667] #0: getpid()
#0: read cover size = 38
#0: getpid = 0xee
completion of call 1 [getpid] on thread 0
out #1: index=1 num=199 errno=0 finished=1 blocked=0 sig=0 cover=38 comps=0
enabling collider
scheduling call 0 [getpid] on thread 0
[   49671] #0: getpid()
#0: read cover size = 38
#0: getpid = 0xee
completion of call 0 [getpid] on thread 0
scheduling call 1 [getpid] on thread 0
worker exiting
unlink(./0/cgroup)
unlink(./0/cgroup.cpu)
unlink(./0/cgroup.net)
rmdir(./0)
```

I think we can get it down to something along the following lines:
```
pid=238 proc=0 threaded=1 collide=1 cover=1 comps=0 dedup=1 fault=0/0/0 prog=0
#0 [15ms] call 0 getpid(...args...)
#0 [25ms] call 0 getpid=0xee errno=0 cover=38 fault=0
#0 [35ms] call 1 getpid(...args...)
#0 [45ms] call 1 getpid=0xee errno=0 cover=38 fault=0
enabling collider
#0 [55ms] call 0 getpid(...args...)
#1 [56ms] call 1 getpid(...args...)
#0 [57ms] call 0 getpid=0xee errno=0 cover=38 fault=0
#1 [58ms] call 1 getpid=0xee errno=0 cover=38 fault=0
done
```

I.e. combine exec opts with pid; 
remove logging of every removed file/dir;
print 1 line when a call starts executing (in the worker thread)
print 1 line when a call finishes (in the worker thread), I guess one generally wants this for race debugging too
print "cover=38" part iff coverage is enabled
print "fault=0" part iff fault injection is enabled for the call
This will lose some info, but that's fine as there is no clear usage story for all of it.

I also think we need to print timestamps from the start of the test process (execute_one already captures start time, we just need to make the var global). I want to see "15ms" and "25ms" rather than do math on "71956997" and "71957003".

If needed later we can introduce a verbose debug output mode. I guess we could just require enabling it in source code rather than treading the flag through layers.

dvyukov commented:
And, yes, debugging and fixing any cases where C repro does not reproduce what syzkaller program triggers would be great!

a13xp0p0v commented:
Thanks, I like your point.
I'll return with the next version (or some additional question first).

a13xp0p0v commented:
Hello Dmitry,

I've updated the commit according to your feedback. Only some minor details
differ. The current version works fine for my needs (race condition debugging).

On 14.09.2018 14:32, Dmitry Vyukov wrote:
> I.e. combine exec opts with pid;

It would cause some intrusive changes with global variables, so I skipped that.
This pid printing doesn't bother me since it happens before the first fuzzing
syscall.

> remove logging of every removed file/dir;

Done.

> print 1 line when a call starts executing (in the worker thread)

Done, but without the syscall number (local var out of scope) and with '->' sign
(to syscall).

> print 1 line when a call finishes (in the worker thread), I guess one generally
> wants this for race debugging too

Agree, done. With '<-' sign (from syscall).

> print "cover=38" part iff coverage is enabled
> print "fault=0" part iff fault injection is enabled for the call

Done.

> This will lose some info, but that's fine as there is no clear usage story for
> all of it.

I've introduced a simple debug_verbose() macro, so the original verbose logs can
be easily switched on in source code.

> I also think we need to print timestamps from the start of the test process
> (execute_one already captures start time, we just need to make the var global).
> I want to see "15ms" and "25ms" rather than do math on "71956997" and "71957003".

Done. But I start counting the time at the beginning of main in executor.cc.
Honestly, I avoid changing the code in those tricky headers.

> If needed later we can introduce a verbose debug output mode. I guess we could
> just require enabling it in source code rather than treading the flag through
> layers.

Yes, done.

Best regards,
Alexander
